Spec JobName for ScrapeConfig CRD

janluak opened the pull request:
prom's scrapeConfig supports providing a specific job-name; scrapeConfigs from CRD no longer bound to `scrapeConfig/<namespace>/<metadata.name>`

Fixes #6147

## Description

Scrape configs from prometheus are supporting a lot more config options compared to the crds aiming to provide the same scrape-config.
With this MR the spec <JobName> is added to the scrape-config-crd. This change makes it possible to specify a name for the scrape-job; otherwise the name defaults to `scrapeConfig/<namespace>/<metadata.name>`

It fixes #6147.

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry


```release-note
feat: support `JobName` spec in ScrapeConfig CRD -> allowing to specify a name for the scrape job instead of default
```

simonpasquier reviewed (commented):
We still need to prevent duplicate job names in the final configuration file otherwise Prometheus will fail to load it.
I wonder if the `jobName` value shouldn't be translated into `metricRelabelConfigs`?

janluak commented:
@simonpasquier is there already existing code for the other crds checking the singularity? Then I'd reuse it...


I tried to write a scrape config with explicitly trying to relabel but I didn't get the scrape_job value to change to the desired name...
If it's the more desired way do you have a relabel config that works? Then I implement it instead...

ArthurSens commented on pkg/prometheus/promcfg.go:
```diff
@@ -2278,7 +2278,13 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 	store *assets.Store,
 ) (yaml.MapSlice, error) {
-	jobName := fmt.Sprintf("scrapeConfig/%s/%s", sc.Namespace, sc.Name)
+	isJobNameSet := sc.Spec.JobName != ""
+	jobName := sc.Spec.JobName
+
+	if !isJobNameSet {
+		jobName = fmt.Sprintf("scrapeConfig/%s/%s", sc.Namespace, sc.Name)
+	}
```
```suggestion
	jobName := fmt.Sprintf("scrapeConfig/%s/%s", sc.Namespace, sc.Name)
	if sc.Spec.JobName != nil {
		jobName = sc.Spec.JobName
	}
```

We can save one memory allocation while improving readability like this :)

janluak replied:
thanks! I reverted the changes due to using the relabel_config now

ArthurSens commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -96,6 +96,8 @@ func (l *ScrapeConfigList) DeepCopyObject() runtime.Object {
 // +k8s:openapi-gen=true
 type ScrapeConfigSpec struct {
+	// The job name assigned to scraped metrics by default.
+	JobName string `json:"jobName,omitempty"`
```
```suggestion
	//+optional
	JobName *string `json:"jobName,omitempty"`
```

We want to follow K8s API conventions for optional fields
(adding optional markers and using pointers)

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

janluak replied:
added, thanks

simonpasquier replied:
@janluak can you apply Arthur's comment? Thanks!

janluak replied:
Sorry, I am a bit confused: the code already contains the `//+optional` comment:
<img width="858" alt="image" src="https://github.com/prometheus-operator/prometheus-operator/assets/12571346/26ab3fcc-d8a8-4e0a-a285-1c157e541b6a">
What did I miss? (see code also [here](https://github.com/janluak/prometheus-operator/blob/feature/crd-scrapeConfig-with-jobName/pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go))

I committed the change myself tough to have the commit message correct: I didn't know how to stay in the structure if using the GitHub-UI to do the commit...

ArthurSens commented:
> @simonpasquier is there already existing code for the other crds checking the singularity? Then I'd reuse it...

Since k8s doesn't allow the same resource kind, with the same name in the same namespace, we never needed to check that. The string formed is always unique; we never needed to check up until now 🤔 .

simonpasquier reviewed (commented):
Regarding the implementation, I wouldn't change the value of `job_name` but rather add a relabeling config to inject the `job` label when `.spec.jobName` is set. This way we wouldn't need to check for `job_name` uniqueness.

janluak commented:
> Regarding the implementation, I wouldn't change the value of `job_name` but rather add a relabeling config to inject the `job` label when `.spec.jobName` is set. This way we wouldn't need to check for `job_name` uniqueness.

sorry for the late reply, got caught up in my project...

Do you have a working relabeling config?
Prior to this MR I tried quite some variants like the one below but after applying it the new label never showed in prometheus. That's why I figured it cannot be overwritten and the `jobName` had to change.

```yaml
  relabelings:
    - targetLabels: ["scrapeJob"]
      replacement: "myName"
      action: replace
```

Do you have a working relabel config? Then I am happy to implement it :)

simonpasquier commented:
In Prometheus config format, it would be:

```
  relabelings:
    - target_label: "job"
      replacement: "myName"
      action: replace
```

janluak commented:
So I played around again with the relabel config and implemented it now with the relabel config.

Still, for some reason I cannot get my head around this issue:
The relabeling worked for two jobs but not for the third one.
<img width="335" alt="image" src="https://github.com/prometheus-operator/prometheus-operator/assets/12571346/6cb348ce-092e-495a-8cf4-99ccdd7f69f5">
<img width="401" alt="image" src="https://github.com/prometheus-operator/prometheus-operator/assets/12571346/1acaa431-2e75-4912-b0f2-d867cac876bb">
<img width="379" alt="image" src="https://github.com/prometheus-operator/prometheus-operator/assets/12571346/6aa4861e-efae-4b2a-b264-9e72c5924e9b">
<img width="366" alt="image" src="https://github.com/prometheus-operator/prometheus-operator/assets/12571346/53253db3-9e1e-4684-aee0-f53336dfbbe6">

First image shows the grafana selection available.
The other three the configs from prometheus config. I would think they are working the same but for some reason the third one is not visible when using the relabel config. When I tested formerly with the fixed setting the name it worked...

Any ideas?

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -2290,6 +2291,15 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 
+	if sc.Spec.JobName != "" {
+		relabelings = append(relabelings, yaml.MapSlice{
+			{Key: "target_label", Value: "scrapeJob"},
```
```suggestion
			{Key: "target_label", Value: "job"},
```

janluak replied:
Could you explain why? I just tried it and now the selector from linkerd-viz can't find the metrics → they are expected at {scrapeJob="linkerd-*"} which isn't fulfilled anymore after this change...

simonpasquier replied:
The `job_name` field from the Prometheus configuration translates to the `job` label. I'm not sure why linkerd-viz would use `scrapeJob` instead.

janluak replied:
hm, what about the label `scrapeJob=scrapeConfig/<namespace>/<metadata.name>`?

If config `job_name` translates to label `job`, would we need another relabel_config here?

simonpasquier replied:
Again I'm not sure where this `scrapeJob` label comes from but it's nothing conventional. If you really need, you can still inject it with another relabel config (e.g. copy from the `job` label).

janluak replied:
oh, sorry, I thought it was the default for `scrapeJob`.

Well, if not I will have to stick with the relabelling on linkerd-viz. I guess I will debug on their side as well why they use it.

I fixed it with #7a2b39b5

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -2290,6 +2291,15 @@ func (cg *ConfigGenerator) generateScrapeConfig(
+			{Key: "action", Value: "replace"},
+			{Key: "replacement", Value: sc.Spec.JobName},
+		})
+		cfg = append(cfg, yaml.MapItem{Key: "relabel_configs", Value: relabelings})
```
```suggestion
```

janluak replied:
If I remove this line (just copied from the `if sc.Spec.RelabelConfigs != nil {` statement) the relabel_config isn't applied anymore. Is there a different trick I missed?

simonpasquier commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -4959,6 +4959,13 @@ func TestScrapeConfigSpecConfig(t *testing.T) {
 			golden: "ScrapeConfigSpecConfig_Empty.golden",
 		},
+		{
+			name: "explicit_job_name",
```
we need another test with a ScrapeConfig spec that defines a non-empty `relabelings` list.

janluak replied:
does commit #ae5a4ad8 suit you? It's not the most elegant solution but works ...

simonpasquier replied:
yes, perfect!

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -2320,6 +2329,9 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		relabelings = append(relabelings, generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.RelabelConfigs))...)
+	}
+
+	if sc.Spec.JobName != "" || sc.Spec.RelabelConfigs != nil {
```
```suggestion
	if len(relabelings) > 0 {
```

janluak replied:
I created two branches:
* feature/crd-scrapeConfig-with-jobName-relabelling-0
* feature/crd-scrapeConfig-with-jobName-relabelling-1

The digit in the end is the same as the comparison used in line 2334 for.

The issue: the suggested refactoring is not equal. 
The variable relabelings is already of `length=1` when initialized.
→ Setting the relabeling to `>0` leaves to a ton of changes of other files not relevant to this feature.
→ Setting the relabeling to `>1` requires only one change to the feature-specific test-file with empty relabel config. 

I'd think the second option is better continuing the current behavior.
Yet, I am familiar enough with the overall behavior to evaluate if the overall change with option one is more desirable...

Just let me know which way you prefer and I merge the one into this merge request branch.
And Merry Christmas ;)

simonpasquier replied:
I see your concern but I think that the current implementation is prone to error. I've submitted #6201 to "fix" it.

janluak replied:
Thanks for your fix. With your refactoring the question of >0 or >1 is obsolete...

ArthurSens reviewed (approved):
Nice solution that you two found!

My initial idea was some to iterate over all selected *Monitor resources and make sure we don't have a matching job name. The relabelings strategy sounds much simpler :)

I think this is good to merge once #6201 gets in, and we add the missing pointer here

simonpasquier commented:
@janluak you'd need to rebase on top of main to resolve the conflict + address @ArthurSens comments.

janluak commented:
> Nice solution that you two found!
> 
> My initial idea was to iterate over all selected *Monitor resources and make sure we don't have a matching job name. The relabelings strategy sounds much simpler :)
> 
> I think this is good to merge once #6201 gets in, and we add the missing pointer here

Thanks for the feedback :)

fixed with #ef52ca90

janluak commented:
> @janluak you'd need to rebase on top of main to resolve the conflict + address @ArthurSens comments.

@simonpasquier thanks for your fix :)

I rebased and implemented the type change as recommended. I hope I didn't screw up the rebase since the label got changed...

ArthurSens commented:
Looks like the rebase was done incorrectly 😅, we don't want to merge 37 unrelated commits

ArthurSens commented:
Hey @janluak , have you noticed the commits that were added by accident here? Do you need some help with the rebase?

janluak commented:
Hi Arthur,Sorry, I was quite busy the last days…Yet, I don’t really have an answer how I revert a rebase. I was thinking of renaming the branch and creating a new one with the same name and then redo the commits but that seems to be quite nasty… other option would be to close this MR in favor of a new one.Do you know the trick how to revert the rebase and then do it correctly?Thanks, JanAm 19.01.2024 um 04:29 schrieb Arthur Silva Sens ***@***.***>:﻿
Hey @janluak , have you noticed the commits that were added by accident here? Do you need some help with the rebase?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ArthurSens commented:
What I usually do is use the git rebase command in interactive mode: `git rebase -i HEAD~<number of commits I want to rebase>`

I believe in the case of this PR you'd work with 37 commits, and you would drop all commits that aren't related to this PR

---

Opening another pull request with only the changes you want also works! Whatever is best for you :)

mviswanathsai commented:
Thanks for the effort @janluak! Do you have the time to fix the rebase?
